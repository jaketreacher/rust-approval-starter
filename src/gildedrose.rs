use std::fmt::{self, Display};
pub struct Item {
    pub name: String,
    pub sell_in: i32,
    pub quality: i32,
}

impl Item {
    pub fn new(name: impl Into<String>, sell_in: i32, quality: i32) -> Item {
        Item {
            name: name.into(),
            sell_in,
            quality,
        }
    }
}

impl Display for Item {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}, {}, {}", self.name, self.sell_in, self.quality)
    }
}

pub struct GildedRose {
    pub items: Vec<Item>,
}

impl Display for GildedRose {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for item in &self.items {
            writeln!(f, "{}", item)?;
        }
        Ok(())
    }
}

impl GildedRose {
    pub fn new(items: Vec<Item>) -> GildedRose {
        GildedRose { items }
    }

    pub fn update_quality(&mut self) {
        for i in 0..self.items.len() {
            if self.items[i].name != "Aged Brie"
                && self.items[i].name != "Backstage passes to a TAFKAL80ETC concert"
            {
                if self.items[i].quality > 0 {
                    if self.items[i].name != "Sulfuras, Hand of Ragnaros" {
                        self.items[i].quality = self.items[i].quality - 1;
                    }
                }
            } else {
                if self.items[i].quality < 50 {
                    self.items[i].quality = self.items[i].quality + 1;

                    if self.items[i].name == "Backstage passes to a TAFKAL80ETC concert" {
                        if self.items[i].sell_in < 11 {
                            if self.items[i].quality < 50 {
                                self.items[i].quality = self.items[i].quality + 1;
                            }
                        }

                        if self.items[i].sell_in < 6 {
                            if self.items[i].quality < 50 {
                                self.items[i].quality = self.items[i].quality + 1;
                            }
                        }
                    }
                }
            }

            if self.items[i].name != "Sulfuras, Hand of Ragnaros" {
                self.items[i].sell_in = self.items[i].sell_in - 1;
            }

            if self.items[i].sell_in < 0 {
                if self.items[i].name != "Aged Brie" {
                    if self.items[i].name != "Backstage passes to a TAFKAL80ETC concert" {
                        if self.items[i].quality > 0 {
                            if self.items[i].name != "Sulfuras, Hand of Ragnaros" {
                                self.items[i].quality = self.items[i].quality - 1;
                            }
                        }
                    } else {
                        self.items[i].quality = self.items[i].quality - self.items[i].quality;
                    }
                } else {
                    if self.items[i].quality < 50 {
                        self.items[i].quality = self.items[i].quality + 1;
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::{GildedRose, Item};

    #[test]
    fn test_acceptance() {
        let names = vec![
            "Aged Brie",
            // "Backstage passes to a TAFKAL80ETC concert",
            // "Sulfuras, Hand of Ragnaros",
            // "Elixir of the Mongoose",
        ];
        let sell_ins = vec![0];
        // let sell_ins = (-1..=12).collect::<Vec<i32>>();
        let qualities = vec![
            // -1,
            0,
            // 1,
            // 49,
            // 50,
            // 51
        ];

        let mut items = vec![];

        for &name in &names {
            for &sell_in in &sell_ins {
                for &quality in &qualities {
                    let item = Item::new(name, sell_in, quality);
                    items.push(item);
                }
            }
        }

        let mut gilded_rose = GildedRose::new(items);
        gilded_rose.update_quality();

        assert_snapshot!(format!("{}", gilded_rose))
    }
}
