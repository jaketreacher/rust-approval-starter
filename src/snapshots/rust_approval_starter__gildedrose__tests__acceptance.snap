---
source: src/gildedrose.rs
expression: "format!(\"{}\", gilded_rose)"
---
Aged Brie, -1, 2

